src/celeritas/ext/detail/GeantGeometryImporter.hh

```diff
@@ -0,0 +1,145 @@
+//----------------------------------*-C++-*----------------------------------//
+// SPDX-FileCopyrightText: 2020 CERN
+// SPDX-License-Identifier: Apache-2.0
+//---------------------------------------------------------------------------//
+/*!
+ * \file GeantGeometryImporter.hh
+ * \brief Class to create a VecGeom model from a pre-existing Geant4 geometry
+ *
+ * Original code from G4VecGeomNav package by John Apostolakis et.al.
+ *
+ * Original source:
+ * https://gitlab.cern.ch/VecGeom/g4vecgeomnav/-/raw/fdd310842fa71c58b3d99646159ef1993a0366b0/include/G4VecGeomConverter.h
+ */
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <utility>
+#include <vector>
+#include <G4PVReplica.hh>
+#include <G4RotationMatrix.hh>
+#include <G4ThreeVector.hh>
+#include <VecGeom/base/TypeMap.h>
+#include <VecGeom/management/GeoManager.h>
+
+class G4LogicalVolume;
+class G4AffineTransformation;
+class G4VPhysicalVolume;
+class G4VSolid;
+
+namespace celeritas
+{
+
+// class converting G4 to VecGeom (only geometry; no materials)
+using vecgeom::BidirectionalTypeMap;
+using vecgeom::LogicalVolume;
+using vecgeom::Transformation3D;
+using vecgeom::VPlacedVolume;
+using vecgeom::VUnplacedVolume;
+
+class GeantGeometryImporter
+{
+  private:
+    /** Remember pointer to generated world from imported G4 geometry. */
+    VPlacedVolume const* world_;
+
+    // one G4 physical volume can correspond to multiple vecgeom placed volumes
+    // (in case of replicas)
+    BidirectionalTypeMap<std::vector<VPlacedVolume const*> const*,
+                         G4VPhysicalVolume const*>
+        placed_volume_map_;
+
+    BidirectionalTypeMap<VUnplacedVolume const*, G4VSolid const*>
+        unplaced_volume_map_;
+
+    BidirectionalTypeMap<LogicalVolume const*, G4LogicalVolume const*>
+        logical_volume_map_;
+
+    // fast O(1) lookup to get VecGeom or G4 placed volume based on index
+    // FastG4VecGeomLookup fFastG4VGLookup;
+    std::vector<Transformation3D const*> replica_transformations_;
+
+    int verbose_;
+
+  public:
+    void set_verbose(int const verbose) { verbose_ = verbose; }
+    int get_verbose_level() const { return verbose_; }
+
+    /// Get placed volume that corresponds to a G4VPhysicalVolume
+    std::vector<VPlacedVolume const*> const*
+    get_placed_volume(G4VPhysicalVolume const* n) const
```

Comment:
Accept and return by const reference. Make the caller check that the volume is non-null instead of doing our own check, unless using pointers substantially simplifies the code.

Reply:
We can leave this as-is if we're moving it to an external repository.